Record input datasets and collections at full parameter path

mvdbeek opened the pull request:
This aligns with most other uses of visit_input_values and fixes the job cache when data inputs are in nested conditionals.  Without the fix the `name` attribute of `JobToInputDatasetAssociation` would not contain the fully prefixed parameter. For trinity
(https://github.com/galaxyproject/tools-iuc/blob/72ae2e05f35098c4cb6dd4f038bff07fd36917ed/tools/trinity/trinity.xml#L195) this would be `pool|left_input`, where `pool` is the outer conditional. With this fix it would be `pool|inputs|left_input`, which matches what we have in the tool state.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

mvdbeek committed:
Record input datasets and collections at full parameter path

This aligns with most other uses of visit_input_values and fixes the job
cache when data inputs are in nested conditionals.  Without the fix the
`name` attribute of `JobToInputDatasetAssociation` would not contain the
fully prefixed parameter. For trinity
(https://github.com/galaxyproject/tools-iuc/blob/72ae2e05f35098c4cb6dd4f038bff07fd36917ed/tools/trinity/trinity.xml#L195)
this would be `pool|left_input`, where `pool` is the outer conditional.
With this fix it would be `pool|inputs|left_input`, which matches what
we have in the tool state.

mvdbeek committed:
Provide legacy unprefixed value handling

mvdbeek committed:
Test ``format_source`` with conditional inputs

mvdbeek committed:
Undo hacks

`LegacyUnprefixedDict` should take care of this

mvdbeek committed:
Implement TreeDict that should handle unprefixed variable access

mvdbeek committed:
Ensure unprefixed PJA rename references continue to work

mvdbeek committed:
Make legacy setting more explicit

jmchilton reviewed (approved):
Looks good to me - sorry about this mess.

mvdbeek committed:
Properly handle overring input parameter in root dict

`TreeDict` will not override ealier set variables, but we do need to
override the catch-all input function if we got an `input` variable
from a conditional or a section.

mvdbeek committed:
Fix unqualified references in ``build_param_dict``

Filters can have `data_ref="some_input" references. If `some_input`
isn't fully qualified this would fail when not using TreeDict.
To avoid having unwrapped references in TreeDict we then rebuild the
TreeDict once more when templating the command line, excluding all
the injected unqualified param_dict references.

We really need to eliminate this, but I think to make this practical for
tool authors we need to build up a tool parameter schema.

mvdbeek committed:
Don't cascade up __current_case__

mvdbeek committed:
Fix input access

mvdbeek committed:
Fix param_dict serialization for data managers

mvdbeek committed:
Try without TreeDict first

Fixes mutect test 3.

mvdbeek committed:
Restrict remap enter function

We're only changing MutableMapping instances, so that's the only items
we need to traverse, I think.
